File: tests/demo_run.cpp
#include "litetorch/tensor.h"
#include "litetorch/ops.h"
#include "litetorch/nn.h"
#include "litetorch/optim.h"
#include "litetorch/data.h"
#include <iostream>
#include <vector>
#include <cmath>
#include <iomanip>
#include <fstream>
#include <sstream>
#include <string>
#include <algorithm>
#include <memory>
#include <cstdio>
#include <sys/resource.h>
#include <chrono>
#include "litetorch/backend.h"

static std::string exec_cmd(const std::string& cmd) {
    std::string out;
    char buffer[256];
    FILE* pipe = popen(cmd.c_str(), "r");
    if (!pipe) return out;
    while (fgets(buffer, sizeof(buffer), pipe)) out += buffer;
    pclose(pipe);
    while (!out.empty() && (out.back() == '\n' || out.back() == '\r' || out.back() == ' ' || out.back() == '\t')) out.pop_back();
    return out;
}

static size_t current_rss_kb() {
    std::ifstream file("/proc/self/status");
    std::string line;
    while (std::getline(file, line)) {
        if (line.rfind("VmRSS:", 0) == 0) {
            std::stringstream ss(line.substr(6));
            size_t kb = 0;
            ss >> kb;
            return kb;
        }
    }
    return 0;
}

static size_t peak_rss_kb() {
    struct rusage usage;
    if (getrusage(RUSAGE_SELF, &usage) == 0) {
        return usage.ru_maxrss;
    }
    return 0;
}

struct GPUInfo {
    bool ok = false;
    double util = 0.0;
    double mem_used = 0.0;
    double mem_total = 0.0;
};

static GPUInfo get_gpu_info() {
    GPUInfo info;
    return info;
}

static double gpu_time_ms() {
    auto backend = litetorch::BackendDispatcher::get().get_backend();
    if (backend && backend->is_available()) {
        backend->finish();
    }
    static auto start_time = std::chrono::high_resolution_clock::now();
    auto now = std::chrono::high_resolution_clock::now();
    return std::chrono::duration<double, std::milli>(now - start_time).count();
}

static int argmax_class(const float* p, int classes) {
    int idx = 0;
    for (int i = 1; i < classes; ++i) {
        if (p[i] > p[idx]) idx = i;
    }
    return idx;
}

static std::pair<double, double> eval_model(std::shared_ptr<litetorch::nn::Sequential> model,
                                            const std::shared_ptr<litetorch::Tensor>& x,
                                            const std::shared_ptr<litetorch::Tensor>& y,
                                            int classes) {
    auto out = model->forward(x);
    auto loss = litetorch::Ops::cross_entropy_loss(out, y);

    const float* op = out->data_ptr();
    const float* yp = y->data_ptr();

    int n = static_cast<int>(x->shape[0]);

    int correct = 0;
    for (int i = 0; i < n; ++i) {
        int pred = argmax_class(op + i * classes, classes);
        int truth = static_cast<int>(yp[i]);
        if (pred == truth) ++correct;
    }

    return std::make_pair(loss->item(), 100.0 * correct / n);
}

int main() {
    std::cout << std::fixed << std::setprecision(6);
    litetorch::manual_seed(13);

    litetorch::Device dev(litetorch::DeviceType::CPU, 0);
    if (litetorch::CLBackend::get().is_available()) {
        dev = litetorch::Device(litetorch::DeviceType::GPU, 0);
    }

    const int classes = 3;
    const int per_class = 200;
    const int n = classes * per_class;
    const float PI = 3.14159265358979323846f;

    std::vector<float> xv;
    std::vector<float> yv;
    xv.reserve(n * 2);
    yv.reserve(n);

    for (int c = 0; c < classes; ++c) {
        for (int i = 0; i < per_class; ++i) {
            float t = static_cast<float>(i) / static_cast<float>(per_class - 1);
            float angle = t * 4.0f * PI + c * 2.0f * PI / classes;
            float radius = 0.20f + 0.80f * t;
            float noise = 0.03f * std::sin(11.0f * t + c * 1.7f);
            float r = radius + noise;
            float x1 = r * std::cos(angle);
            float x2 = r * std::sin(angle);
            xv.push_back(x1);
            xv.push_back(x2);
            yv.push_back(static_cast<float>(c));
        }
    }

    auto x_cpu = litetorch::Tensor::from_vector(xv, {n, 2}, litetorch::Device(litetorch::DeviceType::CPU, 0));
    auto y_cpu = litetorch::Tensor::from_vector(yv, {n}, litetorch::Device(litetorch::DeviceType::CPU, 0));

    std::cout << "--- DATASET ---\n";
    std::cout << "Samples: " << n << " | Classes: " << classes << "\n";
    std::cout << "First 8 points:\n";
    for (int i = 0; i < 8; ++i) {
        std::cout << "(" << x_cpu->data_ptr()[i * 2] << ", " << x_cpu->data_ptr()[i * 2 + 1] << ") ";
    }
    std::cout << "\n\n";

    auto dataset = std::make_shared<litetorch::data::TensorDataset>(x_cpu, y_cpu);
    size_t prefetch_limit = (dev.type == litetorch::DeviceType::GPU) ? 4 : 2;
    litetorch::data::DataLoader loader(dataset, 64, true, dev, prefetch_limit);

    auto model = std::make_shared<litetorch::nn::Sequential>();
    model->add(std::make_shared<litetorch::nn::Linear>(2, 128));
    model->add(std::make_shared<litetorch::nn::LayerNorm>(std::vector<int64_t>{128}));
    model->add(std::make_shared<litetorch::nn::GELU>());
    model->add(std::make_shared<litetorch::nn::Linear>(128, 128));
    model->add(std::make_shared<litetorch::nn::LayerNorm>(std::vector<int64_t>{128}));
    model->add(std::make_shared<litetorch::nn::GELU>());
    model->add(std::make_shared<litetorch::nn::Linear>(128, 64));
    model->add(std::make_shared<litetorch::nn::GELU>());
    model->add(std::make_shared<litetorch::nn::Linear>(64, classes));

    model->to(dev);
    model->compile();

    auto optimizer = std::make_unique<litetorch::optim::Adam>(model->parameters(), 0.01f);
    litetorch::optim::StepLR scheduler(optimizer.get(), 300, 0.8f);

    auto x_eval = x_cpu->to(dev);
    auto y_eval = y_cpu->to(dev);
    std::pair<double, double> init_stats = eval_model(model, x_eval, y_eval, classes);
    std::cout << "--- INITIAL ---\n";
    std::cout << "Loss: " << init_stats.first << " | Accuracy: " << init_stats.second << "%\n";
    std::cout << "RAM: " << current_rss_kb() / 1024.0 << " MB | Peak RAM: " << peak_rss_kb() / 1024.0 << " MB | GPU Time: " << gpu_time_ms() << " ms";

    GPUInfo gi0 = get_gpu_info();
    if (gi0.ok) {
        std::cout << " | GPU: " << gi0.util << "% | VRAM: " << gi0.mem_used << "/" << gi0.mem_total << " MB";
    } else {
        std::cout << " | GPU: N/A";
    }
    std::cout << "\n\n";

    std::cout << "--- TRAINING ---\n";
    for (int epoch = 1; epoch <= 300; ++epoch) {
        loader.reset();
        std::shared_ptr<litetorch::Tensor> batch_x;
        std::shared_ptr<litetorch::Tensor> batch_y;
        while (loader.next(batch_x, batch_y)) {
            optimizer->zero_grad();
            auto out = model->forward(batch_x);
            auto loss = litetorch::Ops::cross_entropy_loss(out, batch_y);
            loss->backward();
            optimizer->step();
        }
        scheduler.step();

        if (epoch == 1 || epoch % 50 == 0) {
            auto x_eval_cur = x_cpu->to(dev);
            auto y_eval_cur = y_cpu->to(dev);
            std::pair<double, double> cur_stats = eval_model(model, x_eval_cur, y_eval_cur, classes);
            std::cout << "Epoch " << epoch
                      << " | Loss: " << cur_stats.first
                      << " | Acc: " << cur_stats.second << "%"
                      << " | LR: " << optimizer->get_lr()
                      << " | RAM: " << current_rss_kb() / 1024.0 << " MB"
                      << " | Peak RAM: " << peak_rss_kb() / 1024.0 << " MB"
                      << " | GPU Time: " << gpu_time_ms() << " ms";
            GPUInfo gi = get_gpu_info();
            if (gi.ok) {
                std::cout << " | GPU: " << gi.util << "% | VRAM: " << gi.mem_used << "/" << gi.mem_total << " MB";
            } else {
                std::cout << " | GPU: N/A";
            }
            std::cout << "\n";
        }
    }

    auto x_final = x_cpu->to(dev);
    auto y_final = y_cpu->to(dev);
    std::pair<double, double> final_stats = eval_model(model, x_final, y_final, classes);
    auto final_out = model->forward(x_final);

    std::cout << "\n--- FINAL PREDICTIONS ---\n";
    for (int i = 0; i < 12; ++i) {
        const float* p = final_out->data_ptr() + i * classes;
        std::cout << "[" << p[0] << ", " << p[1] << ", " << p[2] << "] ";
    }
    std::cout << "\n\n";
    std::cout << "Final Loss: " << final_stats.first << "\n";
    std::cout << "Final Accuracy: " << final_stats.second << "%\n";
    std::cout << "RAM: " << current_rss_kb() / 1024.0 << " MB | Peak RAM: " << peak_rss_kb() / 1024.0 << " MB | GPU Time: " << gpu_time_ms() << " ms";

    GPUInfo gif = get_gpu_info();
    if (gif.ok) {
        std::cout << " | GPU: " << gif.util << "% | VRAM: " << gif.mem_used << "/" << gif.mem_total << " MB";
    } else {
        std::cout << " | GPU: N/A";
    }
    std::cout << "\n";

    return 0;
}

File: tests/demo_run.py
#!/usr/bin/env python3
import litetorch as lt
import time
import os

try:
    import resource
except ImportError:
    resource = None

def current_rss_mb():
    try:
        with open("/proc/self/status") as f:
            for line in f:
                if line.startswith("VmRSS:"):
                    return float(line.split()[1]) / 1024.0
    except Exception:
        pass
    return 0.0

def peak_rss_mb():
    if resource is not None:
        try:
            return float(resource.getrusage(resource.RUSAGE_SELF).ru_maxrss) / 1024.0
        except Exception:
            pass
    return 0.0

def get_vram_mb():
    try:
        out = os.popen("nvidia-smi --query-gpu=memory.used --format=csv,noheader,nounits 2>/dev/null").read().strip()
        if out:
            return float(out.split('\n')[0])
    except Exception:
        pass
    return 0.0

def generate_spiral_data(samples=600, classes=3):
    import math, random
    random.seed(42)
    x_data = []
    y_data = []
    per_class = samples // classes
    pi = math.pi
    for c in range(classes):
        for i in range(per_class):
            t = float(i) / float(per_class - 1)
            angle = t * 4.0 * pi + c * 2.0 * pi / float(classes)
            radius = 0.20 + 0.80 * t
            noise = 0.03 * math.sin(11.0 * t + c * 1.7)
            r = radius + noise
            x1 = r * math.cos(angle)
            x2 = r * math.sin(angle)
            x_data.append(x1)
            x_data.append(x2)
            y_data.append(float(c))
    return x_data, y_data, samples, classes

def argmax(probs, offset, classes):
    best_idx = 0
    best_val = probs[offset]
    for c in range(1, classes):
        val = probs[offset + c]
        if val > best_val:
            best_val = val
            best_idx = c
    return best_idx

def evaluate(model, x, y_data, samples, classes):
    out = model.forward(x)
    probs = out.to_vector()
    correct = 0
    for i in range(samples):
        pred = argmax(probs, i * classes, classes)
        truth = int(y_data[i])
        if pred == truth:
            correct += 1
    accuracy = (correct / float(samples)) * 100.0
    return out, accuracy

def main():
    print("==================================================")
    print("      LITETORCH PYTHON DEMO RUN BENCHMARK")
    print("==================================================")

    x_vec, y_vec, samples, classes = generate_spiral_data(600, 3)
    
    device = lt.auto_device()

    x_tensor = lt.Tensor.from_vector(x_vec, [samples, 2], device, False)
    y_tensor = lt.Tensor.from_vector(y_vec, [samples], device, False)

    l1 = lt.nn.Linear(2, 128, True)
    ln1 = lt.nn.LayerNorm([128])
    l2 = lt.nn.Linear(128, 128, True)
    ln2 = lt.nn.LayerNorm([128])
    l3 = lt.nn.Linear(128, 64, True)
    l4 = lt.nn.Linear(64, classes, True)

    class DeepSpiralNet(lt.nn.Module):
        def __init__(self, l1, ln1, l2, ln2, l3, l4):
            super().__init__()
            self.l1 = l1
            self.ln1 = ln1
            self.l2 = l2
            self.ln2 = ln2
            self.l3 = l3
            self.l4 = l4

        def forward(self, inp):
            h1 = self.l1.forward(inp)
            h1_norm = self.ln1.forward(h1)
            a1 = lt.Ops.gelu(h1_norm)
            h2 = self.l2.forward(a1)
            h2_norm = self.ln2.forward(h2)
            a2 = lt.Ops.gelu(h2_norm)
            h3 = self.l3.forward(a2)
            a3 = lt.Ops.gelu(h3)
            return self.l4.forward(a3)

        def parameters(self):
            return self.l1.parameters() + self.ln1.parameters() + self.l2.parameters() + self.ln2.parameters() + self.l3.parameters() + self.l4.parameters()

    net = DeepSpiralNet(l1, ln1, l2, ln2, l3, l4)
    optimizer = lt.optim.Adam(net.parameters(), lr=0.01)

    start_cpu = time.process_time()
    start_wall = time.perf_counter()

    out, acc = evaluate(net, x_tensor, y_vec, samples, classes)
    loss_init = lt.Ops.cross_entropy_loss(out, y_tensor)

    print(f"Dataset: {samples} samples | Classes: {classes}")
    print(f"Initial Loss: {loss_init.item():.6f} | Initial Accuracy: {acc:.2f}%")
    print(f"Initial RAM: {current_rss_mb():.2f} MB | Peak RAM: {peak_rss_mb():.2f} MB | VRAM: {get_vram_mb():.2f} MB")
    print("--------------------------------------------------")

    for epoch in range(1, 301):
        optimizer.zero_grad()
        out = net.forward(x_tensor)
        loss = lt.Ops.cross_entropy_loss(out, y_tensor)
        loss.backward()
        optimizer.step()

        if epoch % 50 == 0 or epoch == 1:
            out_eval, acc_eval = evaluate(net, x_tensor, y_vec, samples, classes)
            elapsed_cpu_ms = (time.process_time() - start_cpu) * 1000.0
            elapsed_wall_ms = (time.perf_counter() - start_wall) * 1000.0
            print(f"Epoch {epoch:3d} | Loss: {loss.item():.6f} | Acc: {acc_eval:.2f}% | RAM: {current_rss_mb():.2f} MB | Peak RAM: {peak_rss_mb():.2f} MB | VRAM: {get_vram_mb():.2f} MB | CPU Time: {elapsed_cpu_ms:.2f} ms | GPU Time: {elapsed_wall_ms:.2f} ms")

    final_cpu_ms = (time.process_time() - start_cpu) * 1000.0
    final_wall_ms = (time.perf_counter() - start_wall) * 1000.0

    print("--------------------------------------------------")
    print(f"Final Loss: {loss.item():.6f}")
    print(f"Final Accuracy: {acc_eval:.2f}%")
    print(f"Final RAM: {current_rss_mb():.2f} MB | Peak RAM: {peak_rss_mb():.2f} MB | VRAM: {get_vram_mb():.2f} MB")
    print(f"Total CPU Time: {final_cpu_ms:.2f} ms | Total GPU Time: {final_wall_ms:.2f} ms")
    print("==================================================")

if __name__ == "__main__":
    main()
